File: Assets/Script/Data/Protocol/RoleOperation_CreateRoleReturnProto.cs
//===================================================
//作    者：
//创建时间：2022-05-23 20:58:50
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 服务器返回创建角色消息
/// </summary>
public struct RoleOperation_CreateRoleReturnProto : IProtocol
{
    public ushort ProtoCode { get { return 10004; } }

    public bool IsSuccess; //是否成功
    public int MsgCode; //消息码
    public int RoleId; //角色编号
    public byte RoleJob; //角色职业
    public string RoleNickName; //角色昵称
    public int RoleLevel; //角色等级

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteBool(IsSuccess);
            if(IsSuccess)
            {
                ms.WriteInt(RoleId);
                ms.WriteByte(RoleJob);
                ms.WriteUTF8String(RoleNickName);
                ms.WriteInt(RoleLevel);
            }
            else
            {
                ms.WriteInt(MsgCode);
            }
            return ms.ToArray();
        }
    }

    public static RoleOperation_CreateRoleReturnProto GetProto(byte[] buffer)
    {
        RoleOperation_CreateRoleReturnProto proto = new RoleOperation_CreateRoleReturnProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.IsSuccess = ms.ReadBool();
            if(proto.IsSuccess)
            {
                proto.RoleId = ms.ReadInt();
                proto.RoleJob = (byte)ms.ReadByte();
                proto.RoleNickName = ms.ReadUTF8String();
                proto.RoleLevel = ms.ReadInt();
            }
            else
            {
                proto.MsgCode = ms.ReadInt();
            }
        }
        return proto;
    }
}

File: DevelopmentUtility/youyou_Proto/CreateLuaProto/RoleOperation_CreateRoleReturnProto.lua
--服务器返回创建角色消息
RoleOperation_CreateRoleReturnProto = { ProtoCode = 10004, IsSuccess = false, MsgCode = 0, RoleId = 0, RoleJob = 0, RoleNickName = "", RoleLevel = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleOperation_CreateRoleReturnProto.__index = RoleOperation_CreateRoleReturnProto;

function RoleOperation_CreateRoleReturnProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleOperation_CreateRoleReturnProto); --将self的元表设定为Class
    return self;
end


--发送协议
function RoleOperation_CreateRoleReturnProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteBool(proto.IsSuccess);
    if(proto.IsSuccess) then
        ms:WriteInt(RoleId);
        ms:WriteByte(RoleJob);
        ms:WriteUTF8String(RoleNickName);
        ms:WriteInt(RoleLevel);
        else
        ms:WriteInt(MsgCode);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleOperation_CreateRoleReturnProto.GetProto(buffer)

    local proto = RoleOperation_CreateRoleReturnProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.IsSuccess = ms:ReadBool();
    if(proto.IsSuccess) then
        proto.RoleId = ms:ReadInt();
        proto.RoleJob = ms:ReadByte();
        proto.RoleNickName = ms:ReadUTF8String();
        proto.RoleLevel = ms:ReadInt();
        else
        proto.MsgCode = ms:ReadInt();
    end

    ms:Dispose();
    return proto;
end